src/Integration/QualityProfiles/QualityProfileManager.cs

```diff
@@ -0,0 +1,99 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+
+namespace SonarLint.VisualStudio.Integration.RuleSets
+{
+    internal sealed class QualityProfileManager : IDisposable
+    {
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+        private readonly IProjectSystemHelper projectSystemHelper;
+        private readonly IQualityProfileProvider rulesetProvider;
+
+        internal /* for testing purposes */ readonly ConcurrentDictionary<Language, QualityProfile> cachedQualityProfiles =
+            new ConcurrentDictionary<Language, QualityProfile>();
+
+        public QualityProfileManager(IActiveSolutionBoundTracker activeSolutionBoundTracker, IProjectSystemHelper projectSystemHelper,
+            IQualityProfileProvider rulesetProvider)
+        {
+            if (activeSolutionBoundTracker == null)
+            {
+                throw new ArgumentNullException(nameof(activeSolutionBoundTracker));
+            }
+
+            if (projectSystemHelper == null)
+            {
+                throw new ArgumentNullException(nameof(projectSystemHelper));
+            }
+
+            if (rulesetProvider == null)
+            {
+                throw new ArgumentNullException(nameof(rulesetProvider));
+            }
+
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            this.projectSystemHelper = projectSystemHelper;
+            this.rulesetProvider = rulesetProvider;
+
+            activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionBindingChanged;
+
+            if (activeSolutionBoundTracker.IsActiveSolutionBound)
+            {
+                FetchRuleSetForCurrentProjects();
+            }
+
+            // TODO: Find a way to listen to projects added to the solution (in case it is of a language we haven't yet fetched)
+        }
+
+        public void Dispose()
+        {
+            this.activeSolutionBoundTracker.SolutionBindingChanged -= OnSolutionBindingChanged;
+            cachedQualityProfiles.Clear();
+        }
+
+        private void OnSolutionBindingChanged(object sender, ActiveSolutionBindingEventArgs e)
+        {
+            if (!e.IsBound)
+            {
+                cachedQualityProfiles.Clear();
+            }
+            else
+            {
+                FetchRuleSetForCurrentProjects();
+            }
+        }
+
+        private void FetchRuleSetForCurrentProjects()
+        {
+            this.projectSystemHelper.GetFilteredSolutionProjects()
```

Comment:
I think _GetFilteredSolutionProjects_ might expect to be called on the UI thread (which would be another reason for not filtering the projects).